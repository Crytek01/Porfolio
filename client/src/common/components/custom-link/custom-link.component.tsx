import { FunctionComponent } from 'react';

import { ExternalLink, RouterLink } from './custom-link.styles';

export interface ICustomLinkOptions {
  target?: '_self' | '_blank';
  activeColor?: string;
  color?: string;
}

interface ILink {
  to?: string;
  isExternal?: boolean;
  isHoverBoxEnabled?: boolean;
  onClick?: () => void;
  options?: ICustomLinkOptions;
  children: React.ReactNode;
  className?: string;
}

const DEFAULT_OPTIONS = {
  target: '_self',
  color: '#1976d2',
  activeColor: '#6e9cf8;',
};

export const CustomLink: FunctionComponent<ILink> = ({
  isExternal = false,
  isHoverBoxEnabled = true,
  options = {},
  children,
  className,
  to = '#',
  onClick,
}) => {
  const standardOptions = { ...DEFAULT_OPTIONS, ...options };

  const { activeColor, color, ...currentOptions } = standardOptions;

  return isExternal ? (
    <ExternalLink
      className={className}
      href={to}
      onClick={onClick}
      $isHoverBoxEnabled={isHoverBoxEnabled}
      $color={color}
      $activeColor={activeColor}
      {...currentOptions}
    >
      {children}
    </ExternalLink>
  ) : (
    <RouterLink
      className={className}
      to={to}
      onClick={onClick}
      $color={color}
      $activeColor={activeColor}
      $isHoverBoxEnabled={isHoverBoxEnabled}
    >
      {children}
    </RouterLink>
  );
};
