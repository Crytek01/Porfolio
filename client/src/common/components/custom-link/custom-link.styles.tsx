import { Link } from 'react-router-dom';
import styled, { css } from 'styled-components';

interface ILink {
  color: string;
  activeColor: string;
  isHoverBoxEnabled: boolean;
}

const DefaultLinkStyles = css`
  display: inline-block;
  &:link,
  &:visited {
    text-decoration: none;
    color: #1976d2;
    cursor: pointer;

    transition: all 0.3s;
  }

  &:hover,
  &:active {
    color: #6e9cf8;
    outline: none;
  }
`;

export const RouterLink = styled(Link)<ILink>`
  ${DefaultLinkStyles}

  &:link,
  &:visited {
    color: ${({ color }) => color};
  }

  &:hover,
  &:active {
    color: ${({ activeColor }) => activeColor};
    box-shadow: ${({ isHoverBoxEnabled }) =>
      isHoverBoxEnabled ? '0 0 0 0.4rem rgba(0, 204, 255, 0.5)' : 'none'};
  }
`;

export const ExternalLink = styled.a<ILink>`
  ${DefaultLinkStyles}

  &:link,
  &:visited {
    color: ${({ color }) => color};
  }

  &:hover,
  &:active {
    color: ${({ activeColor }) => activeColor};
    box-shadow: ${({ isHoverBoxEnabled }) =>
      isHoverBoxEnabled ? '0 0 0 0.4rem rgba(0, 204, 255, 0.5)' : 'none'};
  }
`;
