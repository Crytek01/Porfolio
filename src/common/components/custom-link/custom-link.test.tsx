import { shallowRender } from '@common/utils/test/shallowRender';
import { describe, expect, it } from 'vitest';

import { CustomLink } from './custom-link.component';

describe('Link component Test', () => {
  it('renders correctly', () => {
    const renderer = shallowRender(<CustomLink>Link</CustomLink>);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
