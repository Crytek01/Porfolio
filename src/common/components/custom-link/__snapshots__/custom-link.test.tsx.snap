// Vitest Snapshot v1

exports[`Link component Test > renders correctly 1`] = `
<Styled(Link)
  $activeColor="#6e9cf8;"
  $color="#1976d2"
  $isHoverBoxEnabled={true}
  to="#"
>
  Link
</Styled(Link)>
`;
